tests/zfs-tests/tests/functional/cli_root/zfs_mount/zfs_mount_test_race.sh

```diff
@@ -0,0 +1,93 @@
+#!/bin/ksh
+
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright (c) 2019 by Tomohiro Kusumi. All rights reserved.
+#
+
+. $STF_SUITE/include/libtest.shlib
+
+#
+# DESCRIPTION:
+# Verify parallel mount ordering is consistent.
+# See github.com/zfsonlinux/zfs/issues/8833 for details.
+#
+# STRATEGY:
+# 1. Create pools and filesystems.
+# 2. Set same mount point for >1 datasets.
+# 3. Unmount all datasets.
+# 4. Mount all datasets.
+# 5. Unmount all datasets.
+# 6. Mount all datasets.
+#
+
+verify_runnable "both"
+
+TMPDIR=${TMPDIR:-$TEST_BASE_DIR}
+MNTPT=$TMPDIR/zfs_mount_test_race_mntpt
+
+DISK1="$TMPDIR/zfs_mount_test_race_disk1"
+TESTPOOL1=zfs_mount_test_race_tp1
+TESTFS1=zfs_mount_test_race_tf1
+
+DISK2="$TMPDIR/zfs_mount_test_race_disk2"
+TESTPOOL2=zfs_mount_test_race_tp2
+TESTFS2=zfs_mount_test_race_tf2
+
+log_must zfs unmount -a # unmount zfs mounts from previous tests
+log_must rm -rf $MNTPT
+log_must rm -rf /$TESTPOOL1
+log_must rm -rf /$TESTPOOL2
+
+function cleanup
+{
+	zpool destroy $TESTPOOL1
+	zpool destroy $TESTPOOL2
+	rm -rf $MNTPT
+	rm -rf /$TESTPOOL1
+	rm -rf /$TESTPOOL2
+	rm -f $DISK1
+	rm -f $DISK2
+	log_must zfs mount -a # remount unmounted zfs mounts
+}
+log_onexit cleanup
+
+log_note "Verify parallel mount ordering is consistent"
+
+log_must truncate -s $MINVDEVSIZE $DISK1
+log_must truncate -s $MINVDEVSIZE $DISK2
+
+log_must zpool create -f $TESTPOOL1 $DISK1
+log_must zpool create -f $TESTPOOL2 $DISK2
+
+log_must zfs create $TESTPOOL1/$TESTFS1
+log_must zfs create $TESTPOOL2/$TESTFS2
+
```

Comment:
I think it would be helpful to add a little ascii art comment here clearly describing the dataset and mountpoint layout being tested.  Are there other interesting/problematic configurations we should be testing?

Reply:
Right, the intention of this layout is not obvious by any means. The only reason I adopted the same layout from the original PR was the race was reproducible. I think there are other layouts that can easily reproduce the race though. I'll add an ascii-art comment or `zfs list` output or something to make it clearer for this layout.

Also note that (while saying canmount property is unrelated to the race) the reason for using `canmount=off` is because this layout fails to mount due to this code below, which is a different topic from the race condition problem. This test itself checks unmount succeeds (i.e. mount order was correct), not mount succeeds.

```
 593         if ((flags & MS_OVERLAY) == 0 && !remount &&
 594             !dir_is_empty(mountpoint)) {
 595                 zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
 596                     "directory is not empty"));
 597                 return (zfs_error_fmt(hdl, EZFS_MOUNTFAILED,
 598                     dgettext(TEXT_DOMAIN, "cannot mount '%s'"), mountpoint));
 599         }
```

Reply:
> because this layout fails to mount due to this code below.

I'm not sure I follow.  Exactly which mount fails in this layout due to overlays not being allowed by default?  (They could be allowed by setting the __overlay__ property).

Reply:
It's in my first comment in #8833,

> The last zfs mount -a failure is caused by /mntpt not being empty for p1/f1 mount (i.e. at this point /mntpt/f2 exists). It has nothing to do with the race.

but actually, since I dropped that "last" mount from `zfs_mount_test_race.sh` (unrelated to test case), I think we can also drop `canmount=off`. canmount does neither good nor bad to this race issue, so it's not a big problem.

Reply:
Sorry, my comment above was wrong. That was when `canmount=off` is used.

When `canmount=off` is unused, `zfs mount -a` in below fails with `cannot mount '/mntpt': directory is not empty`.  canmount does neither good nor bad to this race issue, so I don't see a problem with it even though it's unrelated to the race.


```
log_must ismounted $TESTPOOL1/$TESTFS1
log_must ismounted $TESTPOOL2/$TESTFS2
log_must zfs unmount -a
log_must zfs mount -a /* this fails without `canmount=off` */

log_must ismounted $TESTPOOL1/$TESTFS1
log_must ismounted $TESTPOOL2/$TESTFS2
log_must zfs unmount -a # verify this succeeds
```

Reply:
> I think it would be helpful to add a little ascii art comment here clearly describing the dataset and mountpoint layout being tested.

Added layout AA, and comments on what is expected to happen.